### ISSUE TITLE
Possible memory leak in systemd-localed

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#23777]
**systemd version the issue has been seen with**
 > latest main

**Used distribution**
 > Fedora 35

Discovered with `TEST-73-LOCALE` under ASan+UBSan:

```
Jun 17 22:04:39 H systemd-localed[1124]: =================================================================
Jun 17 22:04:39 H systemd-localed[1124]: ==1124==ERROR: LeakSanitizer: detected memory leaks
Jun 17 22:04:39 H systemd-localed[1124]: Direct leak of 120 byte(s) in 1 object(s) allocated from:
Jun 17 22:04:39 H systemd-localed[1124]:     #0 0x7fb0e87efad7 in calloc (/lib64/libasan.so.6+0xaead7)
Jun 17 22:04:39 H systemd-localed[1124]:     #1 0x40e65c in property_get_locale ../src/locale/localed.c:155
Jun 17 22:04:39 H systemd-localed[1124]:     #2 0x7fb0e7785647 in invoke_property_get ../src/libsystemd/sd-bus/bus-objects.c:448
Jun 17 22:04:39 H systemd-localed[1124]:     #3 0x7fb0e7788ecf in vtable_append_one_property ../src/libsystemd/sd-bus/bus-objects.c:730
Jun 17 22:04:39 H systemd-localed[1124]:     #4 0x7fb0e77895e4 in vtable_append_all_properties ../src/libsystemd/sd-bus/bus-objects.c:786
Jun 17 22:04:39 H systemd-localed[1124]:     #5 0x7fb0e778a068 in property_get_all_callbacks_run ../src/libsystemd/sd-bus/bus-objects.c:846
Jun 17 22:04:39 H systemd-localed[1124]:     #6 0x7fb0e778ef2d in object_find_and_run ../src/libsystemd/sd-bus/bus-objects.c:1356
Jun 17 22:04:39 H systemd-localed[1124]:     #7 0x7fb0e778fe44 in bus_process_object ../src/libsystemd/sd-bus/bus-objects.c:1430
Jun 17 22:04:39 H systemd-localed[1124]:     #8 0x7fb0e77ef0c0 in process_message ../src/libsystemd/sd-bus/sd-bus.c:2960
Jun 17 22:04:39 H systemd-localed[1124]:     #9 0x7fb0e77ef5d7 in process_running ../src/libsystemd/sd-bus/sd-bus.c:3002
Jun 17 22:04:39 H systemd-localed[1124]:     #10 0x7fb0e77f2438 in bus_process_internal ../src/libsystemd/sd-bus/sd-bus.c:3222
Jun 17 22:04:39 H systemd-localed[1124]:     #11 0x7fb0e77f25d2 in sd_bus_process ../src/libsystemd/sd-bus/sd-bus.c:3249
Jun 17 22:04:39 H systemd-localed[1124]:     #12 0x7fb0e77f5cd3 in io_callback ../src/libsystemd/sd-bus/sd-bus.c:3600
Jun 17 22:04:39 H systemd-localed[1124]:     #13 0x7fb0e79ae004 in source_dispatch ../src/libsystemd/sd-event/sd-event.c:3591
Jun 17 22:04:39 H systemd-localed[1124]:     #14 0x7fb0e79b681b in sd_event_dispatch ../src/libsystemd/sd-event/sd-event.c:4175
Jun 17 22:04:39 H systemd-localed[1124]:     #15 0x7fb0e79b797f in sd_event_run ../src/libsystemd/sd-event/sd-event.c:4236
Jun 17 22:04:39 H systemd-localed[1124]:     #16 0x7fb0e7205d94 in bus_event_loop_with_idle ../src/shared/bus-util.c:122
Jun 17 22:04:39 H systemd-localed[1124]:     #17 0x415344 in run ../src/locale/localed.c:819
Jun 17 22:04:39 H systemd-localed[1124]:     #18 0x415619 in main ../src/locale/localed.c:826
Jun 17 22:04:39 H systemd-localed[1124]:     #19 0x7fb0e5cba43f in __libc_start_call_main (/lib64/libc.so.6+0x4043f)
Jun 17 22:04:39 H systemd-localed[1124]: SUMMARY: AddressSanitizer: 120 byte(s) leaked in 1 allocation(s).
```

**Comment by evverx** [comment]
FWIW @mrc0mmand it can be reproduced with `dfuzzer` as well thanks to https://github.com/dbus-fuzzer/dfuzzer/pull/86. I didn't see that memory leak when I tested that feature though so it seems it was introduced relatively recently (probably in 3d36b5d7e7b191fc)

**Comment by mrc0mmand** [comment]
> FWIW @mrc0mmand it can be reproduced with `dfuzzer` as well thanks to [dbus-fuzzer/dfuzzer#86](https://github.com/dbus-fuzzer/dfuzzer/pull/86). I didn't see that memory leak when I tested that feature though so it seems it was introduced relatively recently (probably in [3d36b5d](https://github.com/systemd/systemd/commit/3d36b5d7e7b191fca7c5c65dbab94d99cf5f0230))

Nice! Hopefully, after merging https://github.com/systemd/systemd/pull/23771 I'll test https://github.com/systemd/systemd/pull/22547 once more, and then finally make it run only under sanitizers, so it could be merged as well to avoid issues like this.

**Comment by yuwata** [comment]
Fix is waiting in #23780.
